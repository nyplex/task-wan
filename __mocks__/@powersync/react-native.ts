module.exports = {
  PowerSyncDatabase: jest.fn().mockImplementation(() => ({
    connect: jest.fn(),
    ready: true,
  })),
};
