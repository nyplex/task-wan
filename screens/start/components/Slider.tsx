import React, { useEffect } from "react";
import { useWindowDimensions } from "react-native";
import {
  PanGestureHandler,
  PanGestureHandlerGestureEvent,
} from "react-native-gesture-handler";
import Animated, {
  runOnJS,
  useAnimatedGestureHandler,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from "react-native-reanimated";

type SliderProps = {
  children: React.ReactNode[];
  onIndexChange?: (index: number) => void;
};

const SWIPE_THRESHOLD = 50;

export const Slider: React.FC<SliderProps> = ({ children, onIndexChange }) => {
  const { width } = useWindowDimensions();
  const translateX = useSharedValue(0);
  const currentIndex = useSharedValue(0);

  const onGestureEvent =
    useAnimatedGestureHandler<PanGestureHandlerGestureEvent>({
      onStart: (_, ctx: any) => {
        ctx.startX = translateX.value;
      },
      onActive: (event, ctx: any) => {
        translateX.value = ctx.startX + event.translationX;
      },
      onEnd: (event) => {
        if (
          event.translationX < -SWIPE_THRESHOLD &&
          currentIndex.value < children.length - 1
        ) {
          currentIndex.value += 1;
        } else if (
          event.translationX > SWIPE_THRESHOLD &&
          currentIndex.value > 0
        ) {
          currentIndex.value -= 1;
        }
        if (onIndexChange) {
          runOnJS(onIndexChange)(currentIndex.value);
        }

        translateX.value = withSpring(-currentIndex.value * width, {
          damping: 15,
          stiffness: 150,
        });
      },
    });

  // reset on width change (e.g. rotation)
  useEffect(() => {
    translateX.value = -currentIndex.value * width;
  }, [width]);

  const slidingStyle = useAnimatedStyle(() => ({
    flexDirection: "row",
    transform: [{ translateX: translateX.value }],
  }));

  return (
    <PanGestureHandler onGestureEvent={onGestureEvent}>
      <Animated.View style={{ flex: 1, width, overflow: "hidden" }}>
        <Animated.View style={slidingStyle}>
          {children.map((child, i) => (
            <Animated.View key={i} style={{ width }}>
              {child}
            </Animated.View>
          ))}
        </Animated.View>
      </Animated.View>
    </PanGestureHandler>
  );
};
