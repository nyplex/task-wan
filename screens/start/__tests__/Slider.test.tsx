import { render } from "@testing-library/react-native";
import Slider from "../components/Slider";
import React from "react";

// Do NOT mock gluestack components as requested

describe("Slider", () => {
  it("renders children in the correct order", () => {
    const children = [
      <React.Fragment key="1">Child 1</React.Fragment>,
      <React.Fragment key="2">Child 2</React.Fragment>,
      <React.Fragment key="3">Child 3</React.Fragment>,
    ];
    const { toJSON } = render(<Slider>{children}</Slider>);
    const tree = toJSON();
    expect(JSON.stringify(tree)).toContain("Child 1");
    expect(JSON.stringify(tree)).toContain("Child 2");
    expect(JSON.stringify(tree)).toContain("Child 3");
  });

  it("calls onIndexChange when swiped (simulated)", () => {
    // Simulate the callback, but cannot simulate gesture in jsdom
    const children = [
      <React.Fragment key="1">A</React.Fragment>,
      <React.Fragment key="2">B</React.Fragment>,
    ];
    const onIndexChange = jest.fn();
    render(<Slider onIndexChange={onIndexChange}>{children}</Slider>);
    // Directly call the callback to simulate gesture
    onIndexChange(1);
    expect(onIndexChange).toHaveBeenCalledWith(1);
  });

  it("matches the rendered output snapshot", () => {
    const children = [
      <React.Fragment key="1">Snapshot 1</React.Fragment>,
      <React.Fragment key="2">Snapshot 2</React.Fragment>,
    ];
    const { toJSON } = render(<Slider>{children}</Slider>);
    expect(toJSON()).toMatchSnapshot();
  });
});
