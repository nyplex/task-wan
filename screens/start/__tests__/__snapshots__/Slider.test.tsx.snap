// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider matches the rendered output snapshot 1`] = `
<View
  collapsable={false}
  handlerTag={3}
  handlerType="PanGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  style={
    {
      "flex": 1,
      "overflow": "hidden",
      "width": 750,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "row",
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <View
      style={
        {
          "width": 750,
        }
      }
    >
      Snapshot 1
    </View>
    <View
      style={
        {
          "width": 750,
        }
      }
    >
      Snapshot 2
    </View>
  </View>
</View>
`;
